keps/sig-cli/2229-kubectl-xdg-base-dir/README

```diff
@@ -0,0 +1,195 @@
+| Need to update | test/cmd/legacy-script.sh                                                               | contains reference to $HOME/.kube       |
+| Need to update | test/e2e/kubectl/kubectl.go                                                             | contains reference to $HOME/.kube       |
+| Need to update | test/soak/serve_hostnames/serve_hostnames.go                                            | contains reference to $HOME/.kube       |
+| Need to update | test/e2e/network/scale/localrun/ingress_scale.go                                        | contains reference to $HOME/.kube       |
+| Need to update | test/soak/serve_hostnames/README.md                                                     | contains reference to $HOME/.kube       |
+| Need to update | translations/kubectl/it_IT/LC_MESSAGES/k8s.po                                           | contains reference to $HOME/.kube       |
+| Need to update | translations/kubectl/en_US/LC_MESSAGES/k8s.po                                           | contains reference to $HOME/.kube       |
+| Need to update | translations/kubectl/ja_JP/LC_MESSAGES/k8s.po                                           | contains reference to $HOME/.kube       |
+| Need to update | translations/kubectl/default/LC_MESSAGES/k8s.po                                         | contains reference to $HOME/.kube       |
+| Need to update | translations/kubectl/template.pot                                                       | contains reference to $HOME/.kube       |
+| Need to update | translations/kubectl/de_DE/LC_MESSAGES/k8s.po                                           | contains reference to $HOME/.kube       |
+| Need to update | translations/kubectl/zh_CN/LC_MESSAGES/k8s.po                                           | contains reference to $HOME/.kube       |
+| Need to update | https://github.com/kubernetes-sigs/controller-runtime/blob/cb7f85860a8cde7259b35bb84af1fdcb02c098f2/pkg/client/config/config.go#L129 | Check with project       |
+
+### Test Plan
+
+#### Alpha milestones
+
+Unit tests matching:
+  - update kubectl tests to the new location
+
+#### Beta milestones
+
+Review all unit tests in the main kubernetes tree
+
+#### GA milestones
+All code should be updated with the new config location
+
+### Graduation Criteria
+
+Successful Alpha Criteria
+  - Migrate from $HOME/.kube to $HOME/.config in a transparent way to users
+  - Update Unit tests
+
+#### Alpha -> Beta Graduation
+
+- [x] At least 2 release cycles pass to gather feedback and bug reports during
+  real-world usage
+- [x] End-user documentation is written
+
+#### Beta -> GA Graduation
+
+- [x] At least 2 release cycles pass to gather feedback and bug reports during
+  real-world usage
+- [x] Unit tests must pass and linked in the KEP
+- [x] Documentation exists
+
+### Upgrade / Downgrade Strategy
```

Comment:
This change would break the behavior of kubectl in automation (using config files) as well as the ecosystem of tools that depend on that behavior.  In general this sort of change behooves a significant period of opportunity for deprecation, with some way to opt in to the new behavior.  I’m trying to think of a similar scope of change in behavior we’ve introduced since 1.0 - there haven’t been many, and in many cases we’ve been very conservative.

The most desirable outcome for existing users (and the existing ecosystem of tools around Kube) would probably to make this behavior opt in or configurable at a system level and still allow the existing flow to continue to read files in that location.  Or to attempt to preserve the linkage between the two locations automatically.

I would like to see a more serious set of alternatives considered to measure and mitigate the impact of this change.  It would certainly be dramatic.

Reply:
@smarterclayton Thanks for your input. I have been working to reduce any noise to kubectl users (automation/tools based on kubectl)  via https://github.com/kubernetes/kubernetes/pull/97885 please let me know your thoughts there.

Reply:
I agree that this is a breaking change. Commented on kubernetes/kubernetes#97885

Reply:
any tool that generates / merges a kubeconfig file needs to align on this location or break the UX of the generated config being picked up by default.

tools will have a tough time supporting multiple kubernetes versions. Think cluster creation tools (the first use case to my mind since I maintain one, but there are many of these).

Reply:
Tooling is already broken if it generates into ~/.kube/config and assumes people will automatically read it on the system by that user (the user has "opted out" of using the default location via KUBECONFIG).  Effectively the tool is saying "I know about the public API of the default config file".  If the tool instead used KUBECONFIG location the tool is saying "I know about the public API of the user specifying which config they want, and I am going to deliberately write into that location".  I think both are reasonable - this proposal is impacting the first one, but not the second.

I think the first API is preserved even if you have two locations (X and Y, with X being preferred and the current value), since the API you are targeting is "i know what file is read".  The second API is not broken by any of the alternatives.

Reply:
>  Tooling is already broken if it generates into ~/.kube/config and assumes people will automatically read it on the system by that user (the user has "opted out" of using the default location via KUBECONFIG). 

That is only true when the user has set KUBECONFIG, in my experience many users never set this, in which case tooling should use the default path. We have well defined rules for this [documented](https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/), and tools follow them. 

I don't think it is great to say "we broke this but users just have to start setting KUBECONFIG". As I said above many users especially those just starting out rely on the default location and don't handle setting env across shells well.